File: encodify/Classes/Utilities/Controller/ImageViewController.m
//
//  ImageViewController.m
//  encodify
//
//  Created by 洪朔 on 2016/11/30.
//  Copyright © 2016年 Tuccuay. All rights reserved.
//

#import "ImageViewController.h"

#import "EToast.h"

#import <Masonry/Masonry.h>

#import <Photos/Photos.h>

@interface ImageViewController ()<UIScrollViewDelegate>

@property (strong, nonatomic) UIScrollView *scrollView;
@property (strong, nonatomic) UIImageView *imageView;
@property (strong, nonatomic) UIImage *image;

@end

@implementation ImageViewController

- (instancetype)initWithImage:(UIImage *)image {
    if (self = [super init]) {
        self.image = image;
    }
    
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.view.backgroundColor = [UIColor whiteColor];
    
    self.scrollView = [[UIScrollView alloc] init];
    self.scrollView.delegate = self;
    [self.view addSubview:self.scrollView];
    [self.scrollView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self.view);
    }];
    
    self.imageView = [[UIImageView alloc] initWithImage:self.image];
    [self.scrollView addSubview:self.imageView];
    [self.imageView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self.scrollView);
    }];
    
    self.scrollView.maximumZoomScale = 5;
    CGFloat widthScale = [UIScreen mainScreen].bounds.size.width / self.imageView.bounds.size.width;
    CGFloat navigationBarMaxY = CGRectGetMaxY(self.navigationController.navigationBar.frame);
    CGFloat heightScale = [UIScreen mainScreen].bounds.size.height / (self.imageView.bounds.size.height - navigationBarMaxY);
    CGFloat minScale = MIN(widthScale, heightScale);
    self.scrollView.minimumZoomScale = minScale;
    self.scrollView.zoomScale = minScale;
    
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        dispatch_async(dispatch_get_main_queue(), ^{
            [self updateInsets];
        });
    });
    
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Save" style:UIBarButtonItemStylePlain target:self action:@selector(saveImage)];
}

- (void)saveImage {
    
    PHAuthorizationStatus status = [PHPhotoLibrary authorizationStatus];
    switch (status) {
        case PHAuthorizationStatusNotDetermined: {
            [PHPhotoLibrary requestAuthorization:^(PHAuthorizationStatus status) {
                if (status == PHAuthorizationStatusAuthorized) {
                    [self saveImageToAlbum];
                } else {
                    [self changeSettings];
                }
            }];
            break;
        }
        case PHAuthorizationStatusRestricted: {
            [self changeSettings];
            break;
        }
        case PHAuthorizationStatusDenied: {
            [self changeSettings];
            break;
        }
        case PHAuthorizationStatusAuthorized: {
            [self saveImageToAlbum];
            break;
        }
    }
}

- (void)saveImageToAlbum {
    UIImageWriteToSavedPhotosAlbum(self.image, nil, nil, nil);
    [EToast showStatus:@"Saved"];
}

- (void)changeSettings {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Premission denied" message:@"" preferredStyle:UIAlertControllerStyleAlert];
    [alert addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
    [alert addAction:[UIAlertAction actionWithTitle:@"Settings" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        if ([[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:UIApplicationOpenSettingsURLString]]) {
            [[UIApplication sharedApplication] openURL:[NSURL URLWithString:UIApplicationOpenSettingsURLString]];
        }
    }]];
    [self presentViewController:alert animated:YES completion:nil];
}

- (UIView *)viewForZoomingInScrollView:(UIScrollView *)scrollView {
    return self.imageView;
}

- (void)scrollViewDidZoom:(UIScrollView *)scrollView {
    
    [self updateInsets];
}

- (void)updateInsets {
    
    CGFloat navigationBarMaxY = CGRectGetMaxY(self.navigationController.navigationBar.frame);
    CGFloat insetTop;
    if (self.imageView.frame.size.height > self.scrollView.bounds.size.height - navigationBarMaxY) {
        insetTop = navigationBarMaxY;
    } else {
        insetTop = (self.scrollView.bounds.size.height - self.imageView.frame.size.height) / 2 + navigationBarMaxY;
    }
    
    CGFloat insetLeft = 0.0;
    if (self.imageView.frame.size.width > self.scrollView.bounds.size.width) {
        insetTop = 0;
    } else {
        insetLeft = (self.scrollView.bounds.size.width - self.imageView.frame.size.width) / 2;
    }
    
    UIEdgeInsets inset = UIEdgeInsetsMake(insetTop, insetLeft, 0, 0);
    
    self.scrollView.contentInset = inset;
}

@end

File: encodify/Classes/Utilities/Controllers/ImageViewController.swift
//
//  ImageViewController.swift
//  encodify
//
//  Created by 洪朔 on 2024/12/20.
//  Copyright © 2024年 Tuccuay. All rights reserved.
//

import UIKit
import SnapKit
import Photos

class ImageViewController: UIViewController {
    
    private let image: UIImage
    
    private lazy var scrollView: UIScrollView = {
        let scrollView = UIScrollView()
        scrollView.delegate = self
        scrollView.minimumZoomScale = 0.1
        scrollView.maximumZoomScale = 3.0
        scrollView.showsVerticalScrollIndicator = false
        scrollView.showsHorizontalScrollIndicator = false
        return scrollView
    }()
    
    private lazy var imageView: UIImageView = {
        let imageView = UIImageView(image: image)
        imageView.contentMode = .scaleAspectFit
        return imageView
    }()
    
    init(image: UIImage) {
        self.image = image
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupUI()
        setupConstraints()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        updateZoomScale()
    }
    
    private func setupUI() {
        
        title = "Image Preview"
        
        navigationItem.rightBarButtonItem = UIBarButtonItem(
            title: "Save",
            style: .plain,
            target: self,
            action: #selector(saveImage)
        )
        
        view.addSubview(scrollView)
        scrollView.addSubview(imageView)
    }
    
    private func setupConstraints() {
        scrollView.snp.makeConstraints { make in
            make.edges.equalTo(view.safeAreaLayoutGuide)
        }
        
        imageView.snp.makeConstraints { make in
            make.edges.equalTo(scrollView)
            make.width.height.equalTo(scrollView)
        }
    }
    
    private func updateZoomScale() {
        let scrollViewSize = scrollView.bounds.size
        let imageSize = image.size
        
        let widthScale = scrollViewSize.width / imageSize.width
        let heightScale = scrollViewSize.height / imageSize.height
        let minScale = min(widthScale, heightScale)
        
        scrollView.minimumZoomScale = minScale
        scrollView.zoomScale = minScale
    }
    
    @objc private func saveImage() {
        PHPhotoLibrary.requestAuthorization { [weak self] status in
            DispatchQueue.main.async {
                switch status {
                case .authorized:
                    self?.performSave()
                case .denied, .restricted:
                    self?.showPermissionAlert()
                case .notDetermined:
                    break
                case .limited:
                    self?.performSave()
                @unknown default:
                    break
                }
            }
        }
    }
    
    private func performSave() {
        PHPhotoLibrary.shared().performChanges({
            PHAssetChangeRequest.creationRequestForAsset(from: self.image)
        }) { [weak self] success, error in
            DispatchQueue.main.async {
                if success {
                    Toast.showStatus("Saved to Photos")
                } else {
                    Toast.showError("Save failed")
                }
            }
        }
    }
    
    private func showPermissionAlert() {
        let alert = UIAlertController(
            title: "Photo Library Access Required",
            message: "Please enable photo library access in Settings to save images.",
            preferredStyle: .alert
        )
        
        alert.addAction(UIAlertAction(title: "Settings", style: .default) { _ in
            if let settingsUrl = URL(string: UIApplication.openSettingsURLString) {
                UIApplication.shared.open(settingsUrl)
            }
        })
        
        alert.addAction(UIAlertAction(title: "Cancel", style: .cancel))
        
        present(alert, animated: true)
    }
}

// MARK: - UIScrollViewDelegate
extension ImageViewController: UIScrollViewDelegate {
    func viewForZooming(in scrollView: UIScrollView) -> UIView? {
        return imageView
    }
}
